internal/backend/runtime/omni/controllers/reporting/reporting.go

```diff
@@ -0,0 +1,214 @@
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
+	// Set your Stripe API key from the environment variable.
+	stripe.Key = os.Getenv("STRIPE_API_KEY")
+	if stripe.Key == "" {
+		return fmt.Errorf("STRIPE_API_KEY environment variable not set")
+	}
+
+	// Retrieve Subscription Item ID from environment variables or configuration.
+	subscriptionItemID := os.Getenv("STRIPE_SUBSCRIPTION_ITEM_ID")
+	if subscriptionItemID == "" {
+		return fmt.Errorf("STRIPE_SUBSCRIPTION_ITEM_ID environment variable not set")
+	}
+
+	// Channel to send count changes to the debouncer.
+	countChan := make(chan uint32, 100)
+
+	// Start the debouncer goroutine.
+	panichandler.Go(func() {
+		debouncer(ctx, countChan, subscriptionItemID)
+	}, nil)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+			metrics, err := safe.ReaderGetByID[*omni.MachineStatusMetrics](ctx, r, omni.MachineStatusMetricsID)
+			if err != nil {
+				if state.IsNotFoundError(err) {
+					continue
+				}
+
+				return err
+			}
+
+			count := metrics.TypedSpec().Value.RegisteredMachinesCount
+			select {
+			case <-ctx.Done():
+				return nil
+			case countChan <- count:
+				// Successfully sent to the channel.
+			default:
+				log.Printf("Dropping count update: %d", count)
+			}
+		}
+	}
+}
+
+// debouncer accumulates count changes and updates Stripe after a debounce duration.
+func debouncer(ctx context.Context, countChan <-chan uint32, subscriptionItemID string) {
+	var (
+		pendingCount uint32 = 0
+		timer        *time.Timer
+	)
+
+	// Define a function to process the pending count.
+	processPending := func() {
+		if pendingCount == 0 {
+			return
+		}
+
+		count := pendingCount
+		pendingCount = 0
+
+		// Update the Stripe subscription item with retries.
+		err := updateStripeSubscriptionItemQuantity(subscriptionItemID, count)
+		if err != nil {
+			log.Printf("Failed to update subscription item %s with count %d: %v", subscriptionItemID, count, err)
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			// Context canceled, exit.
+			return
+
+		case count, ok := <-countChan:
+			if !ok {
+				// Channel closed, process any remaining delta and exit.
+				processPending()
+				return
+			}
+
+			pendingCount = count
+
+			// Reset the timer.
+			if timer != nil {
+				if !timer.Stop() {
+					// Drain the timer channel if it has already fired.
+					select {
+					case <-timer.C:
+					default:
+					}
+				}
+			}
+
+			// Start a new timer.
+			timer = time.NewTimer(DebounceDuration)
+
+		case <-func() <-chan time.Time {
+			if timer != nil {
+				return timer.C
+			}
+			// Return a never-blocking channel if timer is nil.
+			return make(chan time.Time)
```

Comment:
reading from `nil` channel blocks forever, so usually it's way easier to write this as `var timerCh <- chan time.Time` and make it non-nil when timer is running